### PR TITLE
Improve `InteractiveTxBuilder` model

### PR DESCRIPTION
**Pull request by pm47**
We introduce a `Purpose` trait, which indicates what the operation is about. Currently there are two possibilities:
- `InitialCommitment`
- `RbfInitialCommitment`

This allows for some nice factoring, but more importantly it will be complemented by a third possibility for splices.

Note that this change preserves the original separation between two main concerns:
- creating the funding tx (in `InteractiveTxParams`)
- creating the commitment (in `InteractiveTxBuilder`).

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala:462]: This makes it clearer what parameters are actually modified.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala:467]: No more repetition (and no more weird `remoteNextCommitInfo.toOption.get`), we explicitly reuse the `common` object that we already have.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:126]: I'm not sure about the name. The real distinction is whether:
- we create the first commitment
- we replace the first commitment
- we spend the existing commitment (splice)

We can make it conceptually independent from the notion of splice. Maybe I'm going too deep.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:137]: Initially I went with:

```suggestion
  case class RbfInitialCommitment(common: Common, commitment: Commitment, previousTransactions: Seq[InteractiveTxBuilder.SignedSharedTransaction]) extends Purpose {
```

It is more pure (there is no ambiguity when rbfing) and works well, but would force to heavily refactor the tests, so I backtracked to get some feedback first.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:126]: It feels like you're trying hard to create an abstraction, but there is a 1:1 mapping between the abstraction and the concrete use-case, so it feels like the abstraction is just making things less clear...What about being more explicit (at least initially) about the scenarios we support?

- `Purpose` can probably stay `Purpose`
- `InitialCommitment` -> `FundingTx`
- `RbfInitialCommitment` -> `FundingTxRbf`
- (future) `ExistingCommitment` -> `SpliceTx`
- (future) `RbfExistingCommitment` -> `SpliceTxRbf`

Those are the only use-cases we can think of right now, and they all need different parameters. We will probably later use `InteractiveTx` for mutual close, but this will also require a slightly different handling, so it will use a different `Purpose`.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:122]: I'm not a big fan of having default values in traits, it's easier to check for correctness when each purpose explicitly sets those values, otherwise a future purpose may incorrectly inherit invalid default values.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:123]: Is this local or remote? I believe it's remote, it's worth being explicit with the name, isn't it?
```suggestion
    def remotePerCommitmentPoint: PublicKey
```

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala:137]: I see what you mean, the `aliceRbf` and `bobRbf` actors that are currently created directly in the fixture are a bit limiting here...it was a bit lazy to implement it this way, it could make sense to refactor this which would unblock more testing scenarios (e.g. more than 1 RBF attempt), and it would make sense to have the `previousTransactions` in this class. I think it would be a nice change, I'll let you decide depending on the amount of work necessary to rework the tests!

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala:462]: But it's also a bit dangerous because we may forget to update some of the parameters...having them all listed ensures we think about each one and whether it can still use the previous value or not. I don't have a strong opinion, both approaches have their pros and cons.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala:467]: I think we should be explicit about using the `metaCommitments` and its first `Commitment` here:

```suggestion
                purpose = InteractiveTxBuilder.RbfInitialCommitment(d.metaCommitments.common, d.metaCommitments.commitments.head),
```

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala:501]:
```suggestion
            purpose = InteractiveTxBuilder.RbfInitialCommitment(d.metaCommitments.common, d.metaCommitments.commitments.head),
```

**Review by t-bast — COMMENTED** [review]
Looks mostly good to me :+1:

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala]: Went with your suggestion in d47aec354.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala]: Renamed in d47aec354.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala]: Done in d47aec354.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala]: Done in d47aec354.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ChannelOpenDualFunded.scala:462]: I prefer the `copy`, as it is more friendly for casual readers. Indeed special care will be needed for future parameters, but I trust our review process!

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala]: Let me take a 2nd shot at it 🤓

Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala]: Refactored tests in 49c763809, and did the change in dbc8b6015. Pretty proud of that one!

**Review by t-bast — APPROVED** [review]
LGTM!

**Pull request merged**
